poc/observe.py

```diff
@@ -0,0 +1,689 @@
+                callback=callback,
+                path=ListenerPath(
+                    node=RequiredTraitListener(
+                        name=name, notify=listener.notify),
+                    nexts=path.nexts
+                ),
+                target=target,
+                dispatcher=dispatcher,
+            )
+
+
+class ListItemListener(BaseListener):
+
+    def __init__(self, notify):
+        self.notify = notify
+
+    def __eq__(self, other):
+        return type(self) is type(other) and self.notify == other.notify
+
+    def prevent_event(self, event):
+        """ Return true if the event should be silenced.
+        """
+        return False
+
+    def create_user_notifier(self, object, callback, target, dispatcher):
+        return TraitObserverNotifier(
+            observer=callback,
+            owner=object._notifiers(True),
+            target=target,
+            dispatcher=dispatcher,
+            prevent_event=self.prevent_event,
+            event_factory=ListObserverEvent,
+        )
+
+    def iter_this_targets(self, object):
+        # object should be a TraitListObject
+        yield object
+
+    def iter_next_targets(self, object):
+        for item in object:
+            if is_notifiable(item):
+                yield item
+
+    def get_change_notifier(self, callback, path, target, dispatcher):
+        return ListenerChangeNotifier(
+            listener_callback=self.change_callback,
+            actual_callback=callback,
+            path=path,
+            target=target,
+            dispatcher=dispatcher,
+            event_factory=ListObserverEvent,
+        )
+
+    @staticmethod
+    def change_callback(event, callback, path, target, dispatcher):
+        logger.debug(
+            "Handling list change. "
+            "({!r})".format(
+                (event.new, event.removed, event.added, path.node))
+        )
+        for item in event.removed:
+            if is_notifiable(item):
+                remove_notifiers(
+                    object=item,
+                    callback=callback,
+                    path=path,
+                    target=target,
+                    dispatcher=dispatcher,
+                )
+        for item in event.added:
+            if is_notifiable(item):
+                add_notifiers(
+                    object=item,
+                    callback=callback,
+                    path=path,
+                    target=target,
+                    dispatcher=dispatcher,
+                )
+
+    def trait_added_matched(self, object, name, trait):
+        """ Return true if an added trait should be handled by this listener
+        """
+        return False
+
+
+class DictValueListener(BaseListener):
+    # TODO: Define this class taking into account
+    # what is going to happen in TraitDictObject
+
+    def __init__(self, notify):
+        self.notify = notify
+
+
+class ListenerPath:
+
+    def __init__(self, node, nexts=()):
+        self.node = node
```

Comment:
What are the types of `node` and `nexts` here, and what do they represent?

Reply:
`node` is an instance of `BaseListener` (described in the main comment).
I removed the examples a few commits ago considering those as noise, but `nexts` are list of `ListenerPath`.

Reply:
Examples were here, not sure if they help or not:
https://github.com/enthought/traits/pull/942/commits/7ff4641e47981554abf0b30844a66de1fe5d8a0b